Migration of order view page printing functionality and layout changes - part 4

matks opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Rebuild of PR https://github.com/PrestaShop/PrestaShop/pull/17617 . When clicked "print" page display its version ready for print. There are some layout changes when print is being initialized.
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | #15819 
| How to test?  | Try clicking print or ctrl + p while in this page **admin-dev/index.php/sell/orders/orders/1/view**. It should be similar to invision sketch of when page is printed. **Dont forget to rebuild assets**

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/17660)
<!-- Reviewable:end -->

matks committed:
Cherry-pick of branch migration/order-view-page/printing-page

matks committed:
Cherry-pick of branch migration/order-view-page/printing-page

Commit:
Fixing print template

Commit:
Fixing print template

matks committed:
Fixing print template - 2

matks committed:
Fixing print template - 2

matks committed:
Update wording for print order

Co-Authored-By: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

matks committed:
Update wording for print order

Co-Authored-By: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

Commit:
Stop displaying address icons and actions buttons

Commit:
Stop displaying address icons and actions buttons

Commit:
Hiding change currency on PDF

Commit:
Hiding change currency on PDF

Commit:
Correcting review on header toolbar and colors

Commit:
Correcting review on header toolbar and colors

Commit:
Revert "Deleting non-used header-toolbar" because header was used in print view

This reverts commit 7491c0c564abf7432573cbf7f501845a659fca2f.

Commit:
Deleting non-used header-toolbar

Commit:
Deleting non-used header-toolbar

matks committed:
Do not override new modals for order view

matks committed:
Do not override new modals for order view

Commit:
Fix print right column

Commit:
Fix print right column

matks committed:
Fix php-cs-fixer

matks committed:
Apply code review feedbacks

matks committed:
Apply code review feedbacks

matks committed:
Fix bad indentation

matks committed:
Fix bad indentation

matks committed:
Fix codestyle

matks committed:
Fix codestyle

matks committed:
Fix codestyle 2

matks committed:
Fix codestyle 2

matks committed:
Fix codestyle 3

matks committed:
Fix codestyle 3

matks committed:
Fix codestyle 4

matks committed:
Fix codestyle 4

Progi1984 commented on admin-dev/themes/new-theme/scss/pages/_orders_view.scss:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    #main-div .content-div {
+      // Forced to use an important there to override inline-style added with JS
```
:+1:

Progi1984 commented on tests/Integration/Behaviour/Features/Context/Domain/OrderRefundFeatureContext.php:
```diff
@@ -35,8 +35,8 @@
 use PrestaShop\PrestaShop\Core\Domain\Order\Command\IssueStandardRefundCommand;
-use PrestaShop\PrestaShop\Core\Domain\Order\Exception\InvalidOrderStateException;
 use PrestaShop\PrestaShop\Core\Domain\Order\Exception\InvalidCancelProductException;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\InvalidOrderStateException;
```
Please rebase your PR : this fix is always fixed in `1.7.7.x` branch.

Progi1984 reviewed (commented):
Need a rebase

eternoendless commented:
Rebased as per @Progi1984 's request

matks committed:
Merge branch 'migration/order-view-page/printing-page-9999-final' of github.com:matks/PrestaShop into migration/order-view-page/printing-page-9999-final

matks committed:
Do not show order messages and alert messages in Print Order

Robin-Fischer-PS commented:
Hi !

Some returns on this PR : 

- Validation/error messages are displayed on the Print, I don't think it should be there
- Message block is also printed, according to the mockup it should not
![image (8)](https://user-images.githubusercontent.com/55688534/74642632-a20be980-5173-11ea-8cf2-35101b8312d8.png)
- If there is more than 8 products, the pagination is printed, it should not (only the first page of the products is printed)
![pr17660 pagination product](https://user-images.githubusercontent.com/55688534/74643115-7e956e80-5174-11ea-968a-300311122598.png)
- "Total", "Validated orders" and "Total spent", and the quantity of a product (if superior to 1) are written is light grey on a white background, it should be in white on dark background.
- For a Pack product, the link "View Pack content" should not be printed"
- The "name" of the document, if saved in PDF, is always the same : Is a "customization" with Order Number possible ?

Here an example of the PDF : 
[Orders • Youpila177 2.pdf](https://github.com/PrestaShop/PrestaShop/files/4213040/Orders.Youpila177.2.pdf)

matks committed:
Fix multiple small issues on print order page

matks commented:
> Hi !
> 
> Some returns on this PR :
> 
> * Validation/error messages are displayed on the Print, I don't think it should be there

Fixed

> * Message block is also printed, according to the mockup it should not

Fixed


> * If there is more than 8 products, the pagination is printed, it should not (only the first page of the products is printed)

Fixed, no pagination and "print page" shows all products

> * "Total", "Validated orders" and "Total spent", and the quantity of a product (if superior to 1) are written is light grey on a white background, it should be in white on dark background.

It seems I am limited by browsers on this.
In Chrome, I cannot control the colors.
In Firefox I can control them a bit.

> * For a Pack product, the link "View Pack content" should not be printed"

Fixed

> * The "name" of the document, if saved in PDF, is always the same : Is a "customization" with Order Number possible ?

It seems I am limited by browsers on this.
In Chrome, I can control the filename. So it's fixed by Chrome.
In Firefox I cannot control it.

matks committed:
Fix multiple small issues on print order page - part 2

matks commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/customer.html.twig:
```diff
@@ -72,14 +72,16 @@
         <p>
-          <span class="badge rounded badge-dark">{{ orderForViewing.customer.validOrdersPlaced }}</span>
+          <span class="badge rounded badge-dark d-print-none">{{ orderForViewing.customer.validOrdersPlaced }}</span>
+          <span class="badge rounded d-none d-print-inline-block">{{ orderForViewing.customer.validOrdersPlaced }}</span>
```
@NeOMakinG I duplicated a lot of items, but each time one has `d-print-none` and the other one has `d-none d-print-...` to have a dedicated display for Print. Is that OK ?

NeOMakinG replied:
Why ? d-print-inline-block is applied only on print screen, it seems useless to do this

matks commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/print_order_statistics.html.twig:
```diff
@@ -12,7 +12,7 @@
         </p>
         <p>
-          <span class="badge rounded badge-dark font-size-100">
+          <span class="badge rounded font-size-100">
```
Did not need to duplicate as this template is for print only

matks commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig:
```diff
@@ -183,7 +183,7 @@
               <div class="col-sm text-center">
                 <p class="text-muted mb-0"><strong>{{ 'Discounts'|trans({}, 'Admin.Global') }}</strong></p>
-                <strong class="orderDiscountsTotal">-{{ orderForViewing.prices.discountsAmountFormatted }}</strong>
+                <strong id="orderDiscountsTotal">-{{ orderForViewing.prices.discountsAmountFormatted }}</strong>
```
Found a mistake on the way 😄 other divs around have IDs, not classes

matks commented:
https://github.com/PrestaShop/PrestaShop/pull/17660/commits/3a60b9548aa351a07a300a6976ebdb89aeb8fe90

> It seems I am limited by browsers on this.
> In Chrome, I cannot control the colors.
> In Firefox I can control them a bit.

Found a workaround  but I need to get @NeOMakinG approval

Otherwise, fixed more feedbacks from @Robin-Fischer-PS  :
- hidden "edit carrier" button for print
- hidden "delete voucher" button for print
- applied my workaround on multiple items that need a white background and black text color for print only

NeOMakinG commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/customer.html.twig:
```diff
@@ -72,14 +72,16 @@
         <p>
-          <span class="badge rounded badge-dark">{{ orderForViewing.customer.totalSpentSinceRegistration }}</span>
+          <span class="badge rounded badge-dark d-print-none">{{ orderForViewing.customer.totalSpentSinceRegistration }}</span>
+          <span class="badge rounded d-none d-print-inline-block">{{ orderForViewing.customer.totalSpentSinceRegistration }}</span>
```
same here

NeOMakinG commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/history.html.twig:
```diff
@@ -28,11 +28,16 @@
           {{ status.name }}
         </span>
+        <span class="badge rounded d-none d-print-inline-block"
+              style="background-color: white; color: black;"
```
Should be inside the stylesheet (I already made a @print screen mediaquery some weeks ago)

NeOMakinG replied:
You could add a class if you need to select it easier

NeOMakinG commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig:
```diff
@@ -226,7 +226,8 @@
               <p class="text-muted mb-0"><strong>{{ 'Total'|trans({}, 'Admin.Global') }}</strong></p>
-              <span class="badge rounded badge-dark font-size-100" id="orderTotal">{{ orderForViewing.prices.totalAmountFormatted }}</span>
+              <span class="badge rounded badge-dark font-size-100 d-print-none" id="orderTotal">{{ orderForViewing.prices.totalAmountFormatted }}</span>
+              <span class="badge rounded font-size-100 d-none d-print-inline-block" id="orderTotal">{{ orderForViewing.prices.totalAmountFormatted }}</span>
```
same as above

matks commented:
So I need to learn CSS 🤔 ...

It was about time !

matks committed:
Reverts workaround from 3a60b9548aa351a07a300a6976ebdb89aeb8fe90 and handles print display properly with media queries

NeOMakinG reviewed (dismissed):
Good job :)

sowbiba commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig:
```diff
@@ -101,11 +101,12 @@
       {% endfor %}
       {% include '@PrestaShop/Admin/Sell/Order/Order/Blocks/View/add_product_row.html.twig' %}
       {% include '@PrestaShop/Admin/Sell/Order/Order/Blocks/View/edit_product_row.html.twig' %}
+
```
empty line not needed

matks replied:
Fixed in https://github.com/PrestaShop/PrestaShop/pull/17660/commits/84aab034cb9252b408a502c914bae5ad65b1a994

Robin-Fischer-PS commented:
Two small bugs, a "normal" one, and unfortunately, a critical one : 

- On Firefox, the pages after the first page are not printed (I can only test the "print in PDF" functionality of the browser, since I do not have a printer at home...) : 
[e marketing module2.pdf](https://github.com/PrestaShop/PrestaShop/files/4215575/e.marketing.module2.pdf)


- For the fix on the "grey" color, could you apply it on Quantity (if superior to 1) and on status order ?
- On voucher table, could you remove the "Actions" column ?
![pr17660 remove column change quantity](https://user-images.githubusercontent.com/55688534/74678821-6b0df600-51bc-11ea-9f14-79997273c0ae.png)
![pr17660 status change](https://user-images.githubusercontent.com/55688534/74678824-6ba68c80-51bc-11ea-86d9-1b235ba3e098.png)

- And last bug, when on Order View, the Product block is large, the "Summary" (Date, Total, Message, Products) which should only appear on the print is displayed under the block : 
![pr17660 summary block](https://user-images.githubusercontent.com/55688534/74678798-59c4e980-51bc-11ea-9ec2-5fc9938c7586.png)

Thanks

matks commented:
> * On Firefox, the pages after the first page are not printed (I can only test the "print in PDF" functionality of the browser, since I do not have a printer at home...) :

On Firefox,
- when using "emulate print" developer tool everything is OK
- when doing "print > open in preview" I see the issue

It seems to be a long known issue with Firefox, see
- https://stackoverflow.com/questions/43492724/firefox-prints-first-page-only-but-other-browsers-doing-fine
- https://stackoverflow.com/questions/13498022/firefox-prints-only-first-page-and-cuts-the-page-on-the-right
- https://stackoverflow.com/questions/45414152/css-flex-box-not-printing-all-pages-on-firefox

@NeOMakinG will try to make it work 🤔 I'll take the small bugs

Robin-Fischer-PS commented:
Sorry for the bad labels, I was really tired yesterday evening :sleepy:

matks committed:
Fix multiple small issues on print order page - part 3

matks commented:
> Two small bugs, a "normal" one, and unfortunately, a critical one :
> 
> * On Firefox, the pages after the first page are not printed (I can only test the "print in PDF" functionality of the browser, since I do not have a printer at home...) :

Bug being explored by @NeOMakinG 

> * For the fix on the "grey" color, could you apply it on Quantity (if superior to 1) and on status order ?

Done

I also tried to remove the black circle but it starts to make things difficult (as most markup is shared between desktop and print) so I kept it like this.

> * On voucher table, could you remove the "Actions" column ?

Done

> * And last bug, when on Order View, the Product block is large, the "Summary" (Date, Total, Message, Products) which should only appear on the print is displayed under the block :

Done

Commit:
Fix print on firefox

Commit:
Fix firefox bug

matks committed:
Fix secondary badge display for print order page

sowbiba commented on admin-dev/themes/new-theme/scss/pages/_orders_view.scss:
```diff
@@ -0,0 +1,98 @@
+
+  > .row:not(.d-print-none) {
+    @media print {
+      display: block !important;
```
This is more a question than a comment

Sometimes you put 
/* stylelint-disable declaration-no-important */
And sometimes not
why ?

Progi1984 replied:
Poke @NeOMakinG

NeOMakinG replied:
Because stylelint is not merged yet.. so anyway I'll have to fix that when I'll rebase it, I think I shouldn't add any stylelint comment at all

sowbiba reviewed (approved):
Approved with a question

Progi1984 commented on admin-dev/themes/new-theme/scss/pages/_orders_view.scss:
```diff
@@ -0,0 +1,98 @@
+      }
+    }
+
+    .card--message {
```
```suggestion
    .card-message {
```

Robin-Fischer-PS commented:
Thanks for the fixes !

Here is the print on Chrome : 
[Order #6 RBWSUZMCF.pdf](https://github.com/PrestaShop/PrestaShop/files/4219756/Order.6.RBWSUZMCF.pdf)

So it's QA :heavy_check_mark: !

matks commented:
Thank you for reviews !
